### DISCUSSION TITLE
Force Renovate to select a node version for postUpgradeTasks

### DISCUSSION THREAD

**Discussion by skutyepov** [renovatebot/renovate#26387] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.105.0

### If you're self-hosting Renovate, select which platform you are using.

github.com

#### Consider the following setup: 

Renovate running with `binarySource="install"` and `allowedPostUpgradeCommands: ["node fail-if-not-node-18.js"]` on a [repository](https://github.com/skutyepov/repo-to-be-renovated/pull/1) with just `package.json` (no package-lock.json) and a post upgrade task to spawn a node process that requires a specific version of node. In the example linked previously, I am running Renovate in containerbase image with node v20 and the minimal repro repository requires node v18 with package.json->engines, .nvmrc, renovate.json->force->constraints->node set to reflect that. 

<details><summary>Dockerfile</summary>

```
FROM containerbase/base:9.30.7
RUN install-tool node 20.10.0
ADD config.js /app/
WORKDIR /app
ENV LOG_LEVEL=debug
CMD npx --yes renovate@37.105.0 --token=...
```

</details>

<details><summary>config.js</summary>

```
module.exports = {
  platform: "github",
  repositories: [
    "skutyepov/repo-to-be-renovated",
  ],
  allowedPostUpgradeCommands: [
    "node fail-if-not-node-18.js",
  ],
  binarySource: "install",
};

```

</details>

#### What actually happens:

Renovate doesn't run `install-tool node v18` because it just updates `package.json` and doesn't need to invoke a package manager to update a lock file. Hence, it remains on node v20 and a post update task fails.

#### Desired behavior: 

If `force.constains.node` is present (and/or some other config option) in `renovate.json`, Renovate switches a node version to satisfy constrains regardless if it's going to invoke a package manager or not. 

### Logs (if relevant)

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

**Comment by rarkins** [comment, marked as answer]
@skutyepov have you tried adding `install-tool node ...` to your post upgrade commands?

**Reply by skutyepov** [reply]
Yes, I have. It installs a specified version of node and the script runs successfully. However, it introduces tight coupling of post upgrade command to containerbase environment. It may not be desirable for some scripts.
IMHO, when Renovate gets enough cues including `force.constrains`, it should invoke `install-tool {$tool} {$version}` like it does when a lock file is present. Or something similar to an already existing option `skipInstalls=false`.

**Reply by rarkins** [reply]
This doesn't make sense to me. You're asking for renovate to install the right node, which means containerbase

**Reply by skutyepov** [reply]
I misunderstood your suggestion, my bad. You meant:

```
"postUpgradeTasks": {
  "commands": ["install-tool node ...", "node post-upgrade-script.js"]
}
```

While, I was thinking about something like `child_process.spawnSync("install-tool node ...");` in _post-upgrade-script.js_ itself 😅 

Thank you for your help and quick feedback @rarkins. I appreciate it. I marked your initial answer as an answer.
